daffodil-core/src/main/scala/org/apache/daffodil/dsom/SequenceGroup.scala

```diff
@@ -147,6 +151,20 @@ abstract class SequenceGroupTermBase(
     }
   }.value
 
+  /**
+   * Provides validation for assert and discriminator placement
+   */
+  protected final lazy val checkIfNonEmptyAndDiscrimsOrAsserts: Unit = {
+      checkDiscrimAndAssertPlacement
+  }
+
+  private lazy val checkDiscrimAndAssertPlacement: Unit = {
+    if (groupMembers.size > 0 && discriminatorStatements.size > 0)
+      SDW(WarnID.DiscouragedDiscriminatorPlacement, "The recommended placement is after the target element.")
+    if (groupMembers.size > 0 && assertStatements.size > 0)
+      SDW(WarnID.DiscouragedAssertPlacement, "The recommended placement is after the target element.")
+  }
+
```

Comment:
I think it's safe to combine these lazy vals. I'm not sure we gain anything by having two separate variables.

I'm also not sure the message makes it clear how to fix this. "After the target element" isn't obvious to me. I'm not sure how to say this with breifly, but maybe something along the lines of

> Discriminators placed directly on sequences has unintutive behavior and will evaluate after the contents of the sequence. Wrap the discriminator in an empty sequence to evaluate before the contents.

That feels horribly long and verbose, but maybe that can be reworded?

Reply:
I'm good with that phrasing. It's only 2 lines of text.